File: ScalaRDF3x/src/main/java/de/mpii/rdf3x/RDFSearch.java
//package de.mpii.rdf3x;
//
//import java.io.IOException;
//import java.sql.ResultSet;
//import java.sql.SQLException;
//import java.util.Properties;
//
//public class RDFSearch {
//
//	public final String rdf3xHome;
//	public final String dbName;
//
//	private Statement s;
//	private Connection c;
//
//	public RDFSearch(String dbName) throws SQLException {
//		this.dbName = dbName;
//        this.rdf3xHome = "rdf3x-0.3.7";
//		Driver d = new Driver();
//		Properties prop = new Properties();
//		prop.put("rdf3xembedded", rdf3xHome + "/bin/rdf3xembedded");
//		this.c = (Connection) d.connect(
//				"rdf3x://" + dbName, prop);
//
//		s = new Statement(c);
//	}
//
//
//    public RDFSearch(String rdf3xHome, String dbName) throws SQLException {
//        this.dbName = dbName;
//        this.rdf3xHome = rdf3xHome;
//        Driver d = new Driver();
//        Properties prop = new Properties();
//        prop.put("rdf3xembedded", rdf3xHome + "/bin/rdf3xembedded");
//        this.c = (Connection) d.connect(
//                "rdf3x://" + dbName, prop);
//
//        s = new Statement(c);
//    }
//
//	public void close() throws SQLException {
//		s.close();
//		c.close();
//	}
//
//
//	/**
//	 * @param args
//	 * @throws java.io.IOException
//	 * @throws InterruptedException
//	 * @throws java.sql.SQLException
//	 */
//	public static void main(String[] args) throws IOException,
//			InterruptedException, SQLException {
//		RDFSearch rdfs = new RDFSearch("/Users/alberto/rdf3x-0.3.7", "/Users/alberto/Documents/Projects/DBpedia_vs_Freebase/DBpedia3.9/rdf3x/dbpedia_properties.rdf3x");
//
//		ResultSet rs = rdfs
//				.SPARQLQuery("select ?name where { ?p <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> ?name } LIMIT 20");
//
//		System.out.println("\nquery executed. results:");
//		while (rs.next()) {
//			System.out.println(rs.getString(1));
//		}
//
//		rs = rdfs
//				.SPARQLQuery("SELECT ?y ?x WHERE { ?y <http://dbpedia.org/ontology/birthDate> ?x . ?z <http://dbpedia.org/ontology/starring> ?y . ?z <http://www.w3.org/2000/01/rdf-schema#label> 'Forrest Gump'}");
//
//		System.out.println("query executed. results:");
//		while (rs.next()) {
//			System.out.println(rs.getString(1));
//		}
//
//		rs.close();
//		rdfs.close();
//	}
//
//	public ResultSet SPARQLQuery(String query) throws SQLException {
//		ResultSet rs = s.executeQuery(query);
//
//
//	}
//
//}

File: ScalaRDF3x/src/main/scala/scalaRdf3x/RDFSearch.scala
package scalaRdf3x

import java.sql.{ResultSet, SQLException}
import java.util.Properties

import de.mpii.rdf3x.{Connection, Driver, Statement}

/**
 * Created by alberto on 7/01/15.
 * Wraps the original RDF3x Java apis.
 */
class RDFSearch(val rdf3xHome: String, val dbName: String) {
	val d: Driver = new Driver
	val prop: Properties = new Properties
	prop.put("rdf3xembedded", rdf3xHome + "/bin/rdf3xembedded")
	val c: Connection = d.connect("rdf3x://" + dbName, prop).asInstanceOf[Connection]
	val s = new Statement(c)


	def this(dbName: String) {
		this("./rdf3x3.7", dbName)
	}


	def close() {
		s.close()
		c.close()
	}

	def SPARQLQuery(query: String) = {
		val rs: ResultSet = s.executeQuery(query)
		val n_variables = rs.getMetaData.getColumnCount
		new Iterator[Map[String, String]] {
			override def next(): Map[String, String] = {
				(for{i <- 1 to n_variables} yield rs.getMetaData.getColumnLabel(i) -> rs.getString(i)).toMap
			}

			override def hasNext() = {
				val hasnext = rs.next()
				if(!hasnext) close()
				hasnext
			}

			private def close() {
				rs.close() }
		}//iterator
	}//SPARQLQuery
}//RDFSearch


object RDFSearch {
	def main(args: Array[String]) {
		val rdf3xhome = args(0)
		val dbPath = args(1)

		val rdfs = new RDFSearch(rdf3xhome, dbPath)
		for (row <- rdfs.SPARQLQuery("select ?name where { <http://dbpedia.org/resource/Tom_Cruise> <http://www.w3.org/2000/01/rdf-schema#label> ?name } LIMIT 20"))
			println(s"${row}")
		rdfs.close()
		println("the end.")
	}//main
}//RDFSearch